text/0114-resolved-ts-for-large-transactions.md

```diff
@@ -0,0 +1,164 @@
+  - Entries should be discarded based on coordinator directives, not due to TTL expiration or cache capacity limits.
+  - Maximize hit rate, i.e. the eviction of an transaction entry should happen after lock release, especially for follower peers.
+
+**Proposed Cache Implementations:**
+1. A large enough capacity to minimize full cache evictions. 
+2. Compact cache entries to minimize memory overhead.
+3. A large enough default TTL.
+4. Post-commit procedure after committing all secondary locks:
+   a. Coordinator broadcasts TTL update to all TiKV nodes.
+   b. Extends TTL by several seconds.
+
+#### 3.2.5 TiKV resolved-ts Resolver
+
+**Proposed Operational Mechanism:**
+1. Standard lock handling remains unchanged.
+2. For large pipelined transaction locks:
+   - Identify by "generation" field.
+   - Track only `start_ts` of locks.
+3. Stop tracking `start_ts` for large pipelined transactions upon LOCK DELETION event.
+
+**Proposed Resolved-ts calculation:**
+- Primary Method:
+  - Map `start_ts` to `min_commit_ts` via txn_status_cache.
+  - Set resolved-ts to `max(min_commit_ts + 1, current_resolved_ts)`.
+- Fallback Method:
+  - Use start_ts if cache lookup fails.
+
+### 3.3 Upgrading TiKV
+
+This proposal is non-intrusive, with automatic fallback to original approach on cache miss, ensuring backward compatibility.
+
+### 3.4 Benefits in resolving locks
+
+Preliminary txn_status_cache lookup can reduce unnecessary overhead for various lock resolution scenarios.
+
+## 4. Compatibility Considerations
+
+Key difference: Services may observe more locks.
+
+**Note:** Current implementation still permits the encounter of locks with timestamps smaller than resolved-ts. This proposal maintains this existing behavior, thus we do not anticipate any correctness issues arising from this modification. The principal challenges we foresee are mainly performance and availability concerns.
+
+### 4.1 Stale read
+
+**Proposed Approach:** Query txn_status_cache first when encountering a lock. Fallback to leader read if not found in cache.
+
+### 4.2 Flashback
+
+1. Compatibility with CDC: Flashback will write a lock to block resolved-ts during its execution. It does not use pipelined transaction so the behavior is not affected.
+2. The current and previous (up to v8.3) implementations of Flashback in TiKV rely on an incorrect assumption about resolved-ts guarantees. This misconception can lead to critical issues, such as the potential violation of transaction atomicity, as documented in https://github.com/tikv/tikv/issues/17415.
+
+### 4.3 EBS snapshot backups
+
+No direct impact. Only dependency on resolved-ts is through Flashback.
+
+### 4.4 CDC
+
+Refactoring work needed, as documented in [Large Transactions Don't Block Watermark](https://github.com/pingcap/tiflow/blob/master/docs/design/2024-01-22-ticdc-large-txn-not-block-wm.md).
+
+## 5. Cost Analysis
+
+- Memory: ~33 bytes per cache entry. TiKV instances can hold millions of entries.
```

Comment:
IS it 33 bytes for each transaction? Why 33 bytes total?

Reply:
Yes. The minimum memory requirement for each cache entry is 33:
8 (start_ts) + 8 (min_commit_ts) + 1 (status) + 8 (TTL) = 33 bytes.
I've clarified this in the doc.